### DISCUSSION TITLE
Redirecting requests based on user's group in NPM

### DISCUSSION THREAD

**Discussion by kestrel-x86** [authelia/authelia#6762] [Q&A]
This might be more of a npm/nginx question.

I have everything working well following the official guide here: https://www.authelia.com/integration/proxies/nginx-proxy-manager/

What I'd like to do is, in certain locations, redirect the request if the authenticated user belongs to a certain group.

I see that the configs provided have a lint `proxy_set_header Remote-Groups $groups;` but I'm not sure exactly what to do with this information. My attempts to access `$groups` results in an empty string.

How can I do this?

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

To achieve redirection based on the authenticated user's group in NGINX (or NGINX Proxy Manager), you need to utilize the `Remote-Groups` header that Authelia sets and passes to NGINX. This header contains the groups the authenticated user belongs to. However, directly using `$groups` in conditions for redirection in NGINX won't work because this variable is not automatically available for use in NGINX's configuration scope. Instead, it's set as a header that can be passed to a proxied application.

To use the `Remote-Groups` information for redirection within NGINX, you'll need to capture this header's value and then use it in an `if` directive or a similar conditional logic block. Unfortunately, NGINX does not natively support directly using HTTP headers in `if` statements for redirection or other logic.

One workaround is to use the `auth_request` module to make a subrequest to a special location that checks the user's group and returns a response based on the group check. Here's a conceptual approach:

1. **Set up an internal location that checks the user's group:**
   - This location will use the `Remote-Groups` header to determine the user's group and return different responses based on the group membership.

2. **Use the `auth_request` directive to make a subrequest to this location:**
   - Based on the response from the subrequest, you can perform redirection or other actions.

Here's an example configuration snippet:

```nginx
# Define an internal location for group checking
location = /check-group {
    internal;
    # Add logic here to check the Remote-Groups header and return appropriate responses
    # For example, return a 403 if the user is not in the required group
    if ($http_remote_groups !~* "required_group") {
        return 403;
    }
    return 204;
}

# Use auth_request in your protected location to make a subrequest to the /check-group location
location /protected-path {
    auth_request /check-group;
    auth_request_set $auth_status $upstream_status;
    
    # Redirect if the subrequest returns 403
    if ($auth_status = 403) {
        return 302 http://example.com/redirect-location;
    }
    
    # Normal proxy_pass configuration here
    proxy_pass http://backend;
}
```

This is a simplified example to illustrate the concept. You'll need to adjust the logic in the `/check-group` location to suit your specific requirements for group checking. Also, remember that using `if` in NGINX requires caution, as it can have unexpected side effects in some contexts.

Unfortunately, due to the limitations of NGINX's configuration language, there isn't a straightforward way to directly use the `Remote-Groups` header in conditional logic for redirection without employing some form of workaround like the one described above.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
Yeah this is probably more a nginx question. I have no idea how to achieve this with nginx and how it handles variable context. I suspect you may be able to use `$upstream_http_remote_groups` or some other variable which references a header set via `proxy_set_header`.

The header is however set with the standard examples you can verify this with a container / app like whoami or podinfo.
